### PR TITLE
drivers: sensor: grow_r502a: Add Missing functionalities and improvements for GROW_R502A fingerprint sensor

### PR DESCRIPTION
**Pull request by DineshDK03**
Add the following functionality and improvements for the R502A fingerprint sensor device.

1. Improvements and bug fixes for the existing mainline version.
2. Add capture, create, load, store, and compare attributes to work with template data from the finger or from the sensor's library index.
3. Add LED control for the R502A sensor's built-in LED.
4. Add upload and download of fingerprint template data.
5. Add Read and write system parameters of the R502A sensor.

### CODE REVIEW

File: drivers/sensor/grow_r502a/grow_r502a.c [hunk 1]
```diff
@@ -287,6 +288,7 @@ static int fps_image_to_char(const struct device *dev, uint8_t char_buf_idx)
 	struct grow_r502a_data *drv_data = dev->data;
 	union r502a_packet rx_packet = {0};
 	char const img_to_char_len = 2;
+	int ret = 0;
```
Comment (MaureenHelm) [inline drivers/sensor/grow_r502a/grow_r502a.c:480]: Are the `ret` changes in this commit related to `drivers: sensor: grow_r502a: move enroll, search, delete to application`? It looks like they may belong in a separate commit

File: drivers/sensor/grow_r502a/grow_r502a.c [hunk 2]
```diff
@@ -461,8 +461,8 @@ static int fps_search(const struct device *dev, uint8_t char_buf_idx)
 		.pid = R502A_COMMAND_PACKET,
 		.data = {R502A_SEARCH, char_buf_idx}
 	};
-	sys_put_be16(R02A_LIBRARY_START_IDX, &tx_packet.data[1]);
-	sys_put_be16(R502A_DEFAULT_CAPACITY, &tx_packet.data[3]);
+	sys_put_be16(R02A_LIBRARY_START_IDX, &tx_packet.data[2]);
+	sys_put_be16(R502A_DEFAULT_CAPACITY, &tx_packet.data[4]);
```
Comment (MaureenHelm) [inline drivers/sensor/grow_r502a/grow_r502a.c:679]: `drivers: sensor: grow_r502a: improvements and bug fixes` should be split up into separate commits for each logical bug fix

File: include/zephyr/drivers/sensor/grow_r502a.h [hunk 3]
```diff
@@ -71,6 +124,12 @@ enum sensor_attribute_grow_r502a {
 	 *	val->val2	matching score.
 	 */
 	SENSOR_ATTR_R502A_COMPARE,
+	/** To control device LED */
+	/**
+	 * @param val->val1	use R502A_LED_PARAMS macro with desired
+	 *			LED configurations.
+	 */
+	SENSOR_ATTR_R502A_DEVICE_LED,
```
Comment (MaureenHelm) [inline include/zephyr/drivers/sensor/grow_r502a.h:169]: Please look at enabling the LED driver API

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: include/zephyr/drivers/sensor/grow_r502a.h (hunk 3 above)
Comment (DineshDK03) [reply include/zephyr/drivers/sensor/grow_r502a.h:169]: LED is internal to the Fingerprint sensor and is accessible to indicate fps actions alone. I don't think we need this LED-controlling functionality to be part of the LED driver. please help me understand why and how? 
  
image

Comment (DineshDK03) [reply include/zephyr/drivers/sensor/grow_r502a.h:169]: @MaureenHelm ping

Comment (DineshDK03) [reply include/zephyr/drivers/sensor/grow_r502a.h:169]: @MaureenHelm ping

Comment (MaureenHelm) [reply include/zephyr/drivers/sensor/grow_r502a.h:169]: It looks like you're enabling functionality similar to the LED API (on, off, blink, breathe). Rather than do it through a custom sensor attribute, I'm suggesting that you implement the LED API for this device

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: include/zephyr/drivers/sensor/grow_r502a.h (hunk 3 above)
Comment (DineshDK03) [reply include/zephyr/drivers/sensor/grow_r502a.h:169]: @MaureenHelm Since this grow_r502a Fingerprint sensor comes under the sensor subsystem and LED is one of the attributes of this grow_r502a sensor, there won't be any need to control this LED as a separate entity, I think it would be best to control this FPS LED with SENSOR_ATTR alone.
If we implement LED API, does it not create any conflict of one device 2 driver API set?

File: include/zephyr/drivers/sensor.h [hunk 4]
```diff
@@ -50,6 +63,8 @@ struct sensor_value {
 	int32_t val1;
 	/** Fractional part of the value (in one-millionth parts). */
 	int32_t val2;
+	/** buffer part of the value*/
+	struct sensor_value_ex ex;
```
[the remark this answers is not in the document]
Comment (DineshDK03) [reply include/zephyr/drivers/sensor.h:67]: It is important for larger chunks of data exported or imported to the driver and introducing ex is inevitable, the footprint is just 8 bytes in a 32-bit machine or shall be made as a confidential compilation.

File: include/zephyr/drivers/sensor/grow_r502a.h (hunk 3 above)
Comment (DineshDK03) [reply include/zephyr/drivers/sensor/grow_r502a.h:169]: @MaureenHelm ping

Comment (DineshDK03) [reply include/zephyr/drivers/sensor/grow_r502a.h:169]: > @MaureenHelm Since this grow_r502a Fingerprint sensor comes under the sensor subsystem and LED is one of the attributes of this grow_r502a sensor, there won't be any need to control this LED as a separate entity, I think it would be best to control this FPS LED with SENSOR_ATTR alone. If we implement LED API, does it not create any conflict of one device 2 driver API set?

@MaureenHelm ping

File: include/zephyr/drivers/sensor.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (DineshDK03) [reply include/zephyr/drivers/sensor.h:67]: @MaureenHelm ping

**Comment by DineshDK03** [comment]
@MaureenHelm please have a look at the comments and do a re-review if needed.

**Comment by DineshDK03** [comment]
@MaureenHelm Please revisit

File: include/zephyr/drivers/sensor.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (teburd) [reply include/zephyr/drivers/sensor.h:67]: Unfortunately I agree with Maureen in this case. I don't think this change can be accepted.

Could the sensor_value two words be used to store the buf pointer and length instead?

Something like...

``` c
struct sensor_value fp_data = {
   .val1 = (int32_t)(buf),
   .val2 = (int32_t)(buf_len),
}
```

I'd note though that passing buffers to/from user mode threads will not work without special care to share the memory from kernel space to user space. So that would be a caveat for this particular driver and use case.

File: include/zephyr/drivers/sensor/grow_r502a.h (hunk 3 above)
Comment (teburd) [reply include/zephyr/drivers/sensor/grow_r502a.h:169]: This can be somewhat resolved by having a single device's data/config structs be shared between two struct device's implementing the two APIs perhaps... but right its a bit confusing how this is actually supposed to work I agree. I think @bjarki-trackunit implemented multiple device APIs for a device before maybe he has some comments on this.

**Review by teburd — CHANGES_REQUESTED** [review]
The sample should be a fingerprint general sample IMO, if generic LED control is desired alongside it DT can be used to connect a fingerprint sensor and a RGB LED

Making them two distinct devices would enable a lot more combinations of this sort of thing than having it all in one sensor device driver.

File: include/zephyr/drivers/sensor.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (yperess) [reply include/zephyr/drivers/sensor.h:67]: I would recommend using the async API instead.

[the remark this answers is not in the document]
Comment (teburd) [reply include/zephyr/drivers/sensor.h:67]: There's an attribute that this seems to be used for

**Review by yperess — CHANGES_REQUESTED** [review]
My suggestion in order to get this PR merged is that since the new data representation is needed for a custom attribute that's declared in the driver's `.h` file, why not just have that attribute be settable via a custom function? It's really no different. So basically users of this driver instead of calling:

```c
sensor_attr_set(
    dev,
    SENSOR_CHAN_FINGERPRINT,
    SENSOR_ATTR_R502A_DOWNLOAD,
    &value
);
```

They would use a specific function like:
```c
grow_r502a_download(dev, buffer, buffer_len);
```

**Comment by teburd** [comment]
sensor wg:
* fingerprint related calls should be driver specific (expose r502a specific functions) as needed
* led if states are internally defined then its ok, if controllable using the API then should be a seperate struct device

**Comment by DineshDK03** [comment]
updated the PR as per the discussions https://discord.com/channels/720317445772017664/1163527713290719373  please have a look and review guys @teburd @yperess

**Review by teburd — DISMISSED** [review]
I'm ok with this approach, needs to pass CI though

File: samples/sensor/grow_r502a/src/main.c [hunk 5]
```diff
@@ -64,13 +119,29 @@ static void trigger_handler(const struct device *dev,
...
+	ret = r502a_read_sys_param(dev,&res);
+	if (ret != 0) {
+		printk("r502a read system parameter failed %d\n", ret);
+		return;
+	}
+
+	printk("baud %d\n", res.baud);
+	printk("addr 0x%x\n", res.addr);
+	printk("lib_size %d\n", res.lib_size);
+	printk("data_pkt_size %d\n", res.data_pkt_size);
```
Comment (kartben) [inline samples/sensor/grow_r502a/src/main.c:140]: not really blocking but I think it would be great to update the contents of https://docs.zephyrproject.org/latest/samples/sensor/grow_r502a/README.html  so that these changes are reflected in the output of the program (there might also be other verbiage worth updating in this page?)

Comment (DineshDK03) [reply samples/sensor/grow_r502a/src/main.c:140]: Updated. Thanks for this comment.

**Review by teburd — APPROVED** [review]

**Review by yperess — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]
